secret_store/src/http_listener.rs

```diff
@@ -0,0 +1,171 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::str::FromStr;
+use std::sync::Arc;
+use hyper::header;
+use hyper::uri::RequestUri;
+use hyper::method::Method as HttpMethod;
+use hyper::status::StatusCode as HttpStatusCode;
+use hyper::server::{Server as HttpServer, Request as HttpRequest, Response as HttpResponse, Handler as HttpHandler,
+	Listening as HttpListening};
+use url::percent_encoding::percent_decode;
+
+use util::ToPretty;
+use super::traits::{Error, ServiceConfiguration};
+use super::key_server::KeyServer;
+use super::{RequestSignature, DocumentAddress, DocumentEncryptedKey};
+
+/// Key server http-requests listener
+pub struct KeyServerHttpListener<T: KeyServer + 'static> {
+	_http_server: HttpListening,
+	handler: Arc<KeyServerSharedHttpHandler<T>>,
+}
+
+/// Parsed http request
+#[derive(Debug, Clone, PartialEq)]
+enum Request {
+	/// Invalid request
+	Invalid,
+	/// Request encryption key of given document for given requestor
+	GetDocumentKey(DocumentAddress, RequestSignature),
+}
+
+/// Cloneable http handler
+struct KeyServerHttpHandler<T: KeyServer + 'static> {
+	handler: Arc<KeyServerSharedHttpHandler<T>>,
+}
+
+/// Shared http handler
+struct KeyServerSharedHttpHandler<T: KeyServer + 'static> {
+	key_server: T,
+}
+
+impl<T> KeyServerHttpListener<T> where T: KeyServer + 'static {
+	/// Start KeyServer http listener
+	pub fn start(config: ServiceConfiguration, key_server: T) -> Result<Self, Error> {
+		let shared_handler = Arc::new(KeyServerSharedHttpHandler {
+			key_server: key_server,
+		});
+		let handler = KeyServerHttpHandler {
+			handler: shared_handler.clone(),
+		};
+
+		let listener_addr: &str = &format!("{}:{}", config.listener_addr, config.listener_port);
+		let http_server = HttpServer::http(&listener_addr).unwrap();
+		let http_server = http_server.handle(handler).unwrap();
+		let listener = KeyServerHttpListener {
+			_http_server: http_server,
+			handler: shared_handler,
+		};
+		Ok(listener)
+	}
+}
+
+impl<T> KeyServer for KeyServerHttpListener<T> where T: KeyServer + 'static {
+	fn document_key(&self, signature: &RequestSignature, document: &DocumentAddress) -> Result<DocumentEncryptedKey, Error> {
+		self.handler.key_server.document_key(signature, document)
+	}
+}
+
+impl<T> HttpHandler for KeyServerHttpHandler<T> where T: KeyServer + 'static {
+	fn handle(&self, req: HttpRequest, mut res: HttpResponse) {
+		if req.method != HttpMethod::Get {
```

Comment:
@rphmeier Yes - we can go that way, but I'm not sure that this would be a part of this particular server/service :) Maybe user/external app will access this keyserver just for storing/retrieving keys && will upload/download files to this `DocumentStore` itself. Another option that this keyserver would be extended to something bigger && will have this functionality itself - then yes, I'll add something like `DocumentStore` here, thanks :)

It will need access to the blockchain data, so it can't be **totally** separate application. But it is intended to be separate executable in future (after IPC will be enabled) - yes.